### DISCUSSION TITLE
Simulation of a capacitor (MLCC) with voltage and Frequencies

### DISCUSSION THREAD

**Discussion by Melopia33** [thliebig/openEMS-Project#266] [General]
Hello,

I have just installed OPEN EMS :-) thank you very much for this great soft , I would like to simulate a capacitor (MLCC) under voltage and frequencies with voltage an frequencies to have impedance, s parameters, strength field , Voltage visualization ... with OPEN EMS. In  the tutorials we can mainly see antenna example but not capacitor. Here's an example of geometry created with OPEN EMS Could you help me please to implement the simulation =>witch excitation and options... should I use ?
Example MLCC @ 500 V and @ 1Ghz

thank you

Xav

![MLCC](https://github.com/user-attachments/assets/7b1d60bf-d8b3-468c-8966-a52c5028aaaa)

**Comment by Melopia33** [comment]
Thank you very much your answer

If I want to try solution number *1* *ideal  lumped capacitance *how can I
manage the simulation with OPEN EMS?
Is it possible to set a voltage with a unique frequency and see the
visualisation of the strength field and voltage inside the geometry?
Which excitation mode should I use to input the voltage on the two metal
terminations?
Sorry for all the questions, I am a chemical Engineer, I know how to
measure capacitors with frequencies but I have no experience in simulation
tools...

best Regards

Xav

**Reply by biergaizi** [reply]
Sorry, I misunderstood your original question. It's why I've deleted it.

I now see that, here, you're in fact simulating a distributed capacitor (with dielectrics, plates, terminals, etc). You're not simulating a lumped capacitor (which is a blackbox without any internal details).

The FDTD simulation workflow is basically the following: FDTD can be understood as virtual VNA or TDT/TDR analyzer. You sent a Gaussian impulse into the Device-Under-Test, and you check the reflection that comes out of it. Then its frequency-domain S-paremeter is calculated automatically (using a Fourier transform) at many different frequencies. Finally you analyze its S-parameter (and optionally the raw E&M field dumps) to understand its frequency response - note that it's up to yourself to interpret the data, openEMS is only a field solver. 

> > Which excitation mode should I use to input the voltage on the two metal terminations?

So conceptually, it should be almost exactly the same as an antenna simulation, you add a thin or zero-thickness Lumped Port that starts from one capacitor terminal, and stops at another capacitor terminal. Use a Gaussian excitation signal, bandwidth 500 - 1 GHz. Run the simulation.

Finally you get the capacitor's wideband S11.

1. If the dielectric you've added is non-linear, using Drude/Lorentz model provided in openEMS (`AddLorentzMaterial()` with the correct physical parameters), the S-parameter should show frequency-dependent capacitance (after interpretation). 
2. If the dielectric you've added is linear (`AddMaterial()` with a single value of permittivity).  S11 may still go up and down show resonances at many different frequencies - when the size of the capacitor is comparable to the signal wavelength, you should see transmission-line behaviors such as quarter-wave transformers.

Optionally you can enable "field dump", which periodically writes raw E&M field data on disk. This can be visualized by the general-purpose data analysis tool ParaView.

Thus, I suggest you to try this option first. 

However, this is only the theory... In practice many subtle details can go wrong, and the results often can be unphysical without any obvious cause. If it happens, the best hope is posting the sample code here and see if someone can help you (In fact, I'm writing a tutorial on capacitor modeling, I find even the simplest example, the parallel-plate capacitor/waveguide, is not that straightforward due to these problems)...

> Is it possible to set a voltage with a unique frequency and see the visualisation of the strength field and voltage inside the geometry?

Yes, instead of a Gaussian excitation, you can also use a Sinusoidal excitation (`SetSinusExcite()`). The FDTD simulation will keep running until steady-state is detected. This mode is rarely used since most FDTD simulations are wideband simulations, but it would be useful if you want to take a look at the steady-state field strength inside the capacitor. 

The Parallel Plate Waveguide example is the simplest demo that uses this feature:

Tutorial (slightly outdated): https://wiki.openems.de/index.php/Tutorial:_Parallel_Plate_Waveguide.html
Code: https://github.com/thliebig/openEMS/blob/f84be9f817e8cd5fc6faec1dd078c0fb101ca013/matlab/Tutorials/Parallel_Plate_Waveguide.m#L20

**Reply by Melopia33** [reply]
Thank you very much I will try in this way

**Reply by Melopia33** [reply]
Hello,


Is it to possible to check please my code for the simple Following model (I try to folow your precious advice) ?
The first goal is to see the E field inside the capacitor with volltage
Later I will try S paramaters also  with your advice

Please find the main steps

Goemetric model:

![image](https://github.com/user-attachments/assets/9224470a-19ee-411a-a4a1-d4dcec9ec4f7)
X is the direction of the lenght of the electrode
Y is the direction of the width of the electrode
unit = 1e-6

% init and define FDTD parameter F=10 Mhz

FDTD = InitFDTD(100,0,'OverSampling',50);
FDTD = SetSinusExcite(FDTD,10e6);
BC = {'PEC' 'PEC' 'PMC' 'PMC' 'MUR' 'MUR'}
FDTD = SetBoundaryCond(FDTD,BC);

For the mesh I choose a simple mesh (must be improved after)

mesh.x = SmoothMeshLines([0 layerLength], 1)
mesh.y = SmoothMeshLines([0 layerWidth], 1)
mesh.z = SmoothMeshLines([-layerThickness zPos+layerThickness], 1)
CSX = DefineRectGrid(CSX, 1e-6, mesh);

% define the excitation The E=10 V/m in Z direction (unit =1e-6) =>the exitation is in the  plane yz (termination), I am not sure of that..
CSX = AddExcitation(CSX,'excitation',0,[0 0 10]);
CSX = AddBox(CSX,'excitation',0,[0 0 0],[0, layerWidth,zPos+layerThickness]); 

![image](https://github.com/user-attachments/assets/2705ff48-d47e-431a-aad2-3bdfa575d34c)

% define a time domain e-field dump box in the xz plane
CSX = AddDump(CSX,'Et','DumpMode',0);
CSX = AddBox(CSX,'Et',0,[0,layerWidth/2,0],[layerLength,layerWidth/2,zPos+met_layerThickness]);

![image](https://github.com/user-attachments/assets/9a49b3b0-aa1f-45f7-81ee-4870de99ed4f)


Best Regards

Xav

**Reply by Melopia33** [reply]
After all, for the excitation I have put two excitations to setup the 2 voltages + and  - (E n the Z direction)

% define the excitation electrode 1
CSX = AddExcitation(CSX,'excitationp1',0,[0 0 10E6]);
CSX = AddBox(CSX,'excitationp1',0,[0 0 layerThickness],[layerLength - met_layerThickness - met_gap, layerWidth,layerThickness]);

% define the excitation electrode 2
CSX = AddExcitation(CSX,'excitationm2',0,[0 0 -10E6]);
CSX = AddBox(CSX,'excitationm2',0,[met_layerThickness + met_gap 0 layerThickness + 2*met_layerThickness + layerThickness],[layerLength, layerWidth,layerThickness + 2*met_layerThickness + layerThickness]);

Visualisation of the 2 excitations
![Excitation](https://github.com/user-attachments/assets/49ca4633-1a05-46b9-9f15-5dce0006ad03)

When I try to visualize the Field in the x, z with paraview I can't see anything (the field is extremely weak).
![paraview](https://github.com/user-attachments/assets/cb78ac48-6649-4eab-9f78-6135e49396ac)


So I think my code isn't correct, could you please help me?

When I use paraview with the tutorial example  Parallel Plate Waveguide it's OK 
![Plate](https://github.com/user-attachments/assets/0c76947c-3a91-4001-8882-4a0bfda7838b)

**Reply by VolkerMuehlhaus** [reply]
> I have put two excitations to setup the 2 voltages + and - (E n the Z direction)

I don't understand what you are trying to do here. It seems that you misunderstand the port as a potential, but it's really a voltage across two terminals! 

I would define one lumped port between the electrodes. That lumped port has two terminals, one goes to each electrode.

**Reply by Melopia33** [reply]
Thank you very much if I understand I must use [CSX, port{1}] = AddLumpedPort for each electrode but how can I manage the sinus excitation voltage between the 2 electrodes with AddExcitation and FDTD = SetSinusExcite(FDTD,10e6);!,

**Reply by VolkerMuehlhaus** [reply]
> if I understand I must use [CSX, port{1}] = AddLumpedPort for each electrode

Yes, you should use lumped port for this..

Each port is defined between two conductors. If you have only two electrodes, with no other common ground, you only need ONE port BETWEEN the electrodes! 

If you have two electrodes plus common ground, as shown in your initial question, you need TWO ports: one from the left electrode to ground, the other from the right electrode to ground. 

Can you upload your latest model code and the resulting XML for AppCSXAD, so that we can see what you modelled?

**Reply by VolkerMuehlhaus** [reply]
To avoid misunderstanding: use lumped port, don't put excitation into your code.

The lumped port has everything inside that you need: the source with an internal resistor and the voltage & current probe.

**Reply by Melopia33** [reply]
I have tried to followed your advice, I have added the two ports (one for each termination) 

startPort1 = [-1.0, 0.0, 0.0];
stopPort1 = [met_layerThickness/2, layerWidth, -(zPos + layerThickness + met_layerThickness)*0.2+1];
[CSX, port{1}] = AddLumpedPort(CSX, 5, 1, 50, startPort1, stopPort1, [0 0 1], true);

startPort2 = [layerLength-1.0, 0, 0];
stopPort2 = [layerLength+met_layerThickness/2, layerWidth, -(zPos + layerThickness + met_layerThickness)*0.2+1];
[CSX, port{2}] = AddLumpedPort(CSX, 5, 2, 50, startPort2, stopPort2, [0 0 1], true)

![modèle](https://github.com/user-attachments/assets/1b62d5c3-9f4a-4399-bc85-7a989829c9dd)

For the excitation I've used SetSinusExcite(FDTD,10e6) and for  AddExcitation I put the volume of the dielectric, where there is the Voltage :  is it correct?

% define the excitation in the dielectric
CSX = AddExcitation(CSX,'excitationp1',0,[0 0 10E6]);
CSX = AddBox(CSX,'excitationp1',0,[met_layerThickness, 0, met_layerThickness+layerThickness],[layerLength-met_layerThickness, layerWidth, layerThickness + met_layerThickness+layerThickness]);


Visualisation of the 2 ports:
![ports](https://github.com/user-attachments/assets/cbd51398-9a78-4c5a-99a9-7b140b08c279)

Visualisation of the excitation volume:
![Excitation volume](https://github.com/user-attachments/assets/b41067bb-3abb-4501-8f59-036ea0e70bcd)

**Reply by VolkerMuehlhaus** [reply]
No, you did not follow my advice. It seems that you **totally** misunderstand the use of ports in openEMS.

Don't put in additional excitation in your code - that is already included in the port!
And don't forget that current only flows in closed loops - you forgot the common ground (=return path) at the bottom of your ports that allows current flow from left to right.

I can only run Python models here, so unfortunately I can't create a working Matlab model for you. 
But I tried to sketch two alternative models, as 1-port model and 2-port model

<img width="543" alt="cap2" src="https://github.com/user-attachments/assets/900cf633-5f78-45ff-b4af-0419f266b015" />

**Reply by Melopia33** [reply]
Thank you again for your answer and your patience !!!
I have choosen the 2 port model:

![modèle 2 ports](https://github.com/user-attachments/assets/4de299ef-13a2-44f2-8079-d24ebfc5fe85)

I have removed the additional excitation:

For the 2 ports with 500 V port 1 and port 2 are in the Z direction [0,0,1]

startPort1 = [-1.0, 0.0, 0.0];
stopPort1 = [met_layerThickness/2, layerWidth, -(zPos + layerThickness + met_layerThickness)*0.2+1];
[CSX, port{1}] = AddLumpedPort(CSX, 5, 1, 50, startPort1, stopPort1, [0 0 1], true,500);

startPort2 = [layerLength-1.0, 0, 0];
stopPort2 = [layerLength+met_layerThickness/2, layerWidth, -(zPos + layerThickness + met_layerThickness)*0.2+1];
[CSX, port{2}] = AddLumpedPort(CSX, 5, 2, 50, startPort2, stopPort2, [0 0 1], true,0);

For the ground: 

start = [-layerLength*0.1, -layerWidth*0.1, -(zPos + layerThickness + met_layerThickness)*0.2];
stop = [layerLength*1.1, layerWidth*1.1, -(zPos + layerThickness + met_layerThickness)*0.2+1];  
CSX = AddBox(CSX, 'metal', 1, start, stop);


Is it better now?
